File: keyboards/ios/LisanUdDawatKeyboard/KeyboardColors.swift
import UIKit

// Centralised adaptive colour palette.
// Every UIColor here resolves automatically when the trait collection changes —
// no traitCollectionDidChange overrides required for backgroundColor.
// Core Graphics drawing (KeyCalloutView) must call .resolvedColor(with: traitCollection).

enum KeyboardColors {

    // MARK: - Backgrounds

    static let background = UIColor { tc in
        tc.userInterfaceStyle == .dark
            ? UIColor(red: 0.137, green: 0.137, blue: 0.145, alpha: 1)   // #232325
            : UIColor(red: 0.812, green: 0.820, blue: 0.843, alpha: 1)   // #CFD1D7
    }

    static let predictiveBar = UIColor { tc in
        tc.userInterfaceStyle == .dark
            ? UIColor(red: 0.110, green: 0.110, blue: 0.118, alpha: 1)   // #1C1C1E
            : UIColor(red: 0.839, green: 0.847, blue: 0.863, alpha: 1)   // #D6D8DC
    }

    static let separator = UIColor { tc in
        tc.userInterfaceStyle == .dark
            ? UIColor(white: 0.28, alpha: 1)
            : UIColor(white: 0.63, alpha: 1)
    }

    // MARK: - Key backgrounds

    static let characterKey = UIColor { tc in
        tc.userInterfaceStyle == .dark
            ? UIColor(red: 0.420, green: 0.420, blue: 0.439, alpha: 1)   // #6B6B70
            : .white
    }

    static let specialKey = UIColor { tc in
        tc.userInterfaceStyle == .dark
            ? UIColor(red: 0.165, green: 0.165, blue: 0.176, alpha: 1)   // #2A2A2D
            : UIColor(red: 0.651, green: 0.675, blue: 0.718, alpha: 1)   // #A6ACB7
    }

    // Pressed state: character keys darken to match special key hue
    static let pressedKey = UIColor { tc in
        tc.userInterfaceStyle == .dark
            ? UIColor(red: 0.239, green: 0.239, blue: 0.255, alpha: 1)
            : UIColor(red: 0.651, green: 0.675, blue: 0.718, alpha: 1)
    }

    // MARK: - Shift locked state (inverted for emphasis)

    static let shiftLockedBackground = UIColor { tc in
        tc.userInterfaceStyle == .dark
            ? UIColor(white: 0.85, alpha: 1)
            : UIColor(white: 0.22, alpha: 1)
    }

    static let shiftLockedText = UIColor { tc in
        tc.userInterfaceStyle == .dark ? .black : .white
    }

    // MARK: - Callout & popup

    static let calloutBubble = UIColor { tc in
        tc.userInterfaceStyle == .dark
            ? UIColor(red: 0.310, green: 0.310, blue: 0.325, alpha: 1)
            : UIColor(white: 0.20, alpha: 1)
    }

    static let popup = UIColor { tc in
        tc.userInterfaceStyle == .dark
            ? UIColor(red: 0.255, green: 0.255, blue: 0.267, alpha: 1)
            : UIColor(white: 0.96, alpha: 1)
    }
}

File: keyboards/android/app/src/main/java/com/lisanuddawat/keyboard/KeyboardColors.kt
package com.lisanuddawat.keyboard

import android.content.Context
import android.content.res.Configuration
import android.graphics.Color

object KeyboardColors {

    fun background(ctx: Context)      = if (dark(ctx)) Color.parseColor("#232325") else Color.parseColor("#CFD1D7")
    fun predictiveBar(ctx: Context)   = if (dark(ctx)) Color.parseColor("#1C1C1E") else Color.parseColor("#D6D8DC")
    fun separator(ctx: Context)       = if (dark(ctx)) Color.parseColor("#474747") else Color.parseColor("#A1A1A1")
    fun characterKey(ctx: Context)    = if (dark(ctx)) Color.parseColor("#6B6B70") else Color.WHITE
    fun specialKey(ctx: Context)      = if (dark(ctx)) Color.parseColor("#2A2A2D") else Color.parseColor("#A6ACB7")
    fun pressedKey(ctx: Context)      = if (dark(ctx)) Color.parseColor("#3D3D41") else Color.parseColor("#A6ACB7")
    fun shiftLockedBg(ctx: Context)   = if (dark(ctx)) Color.parseColor("#D9D9D9") else Color.parseColor("#383838")
    fun shiftLockedText(ctx: Context) = if (dark(ctx)) Color.BLACK               else Color.WHITE
    fun keyLabel(ctx: Context)        = if (dark(ctx)) Color.WHITE                else Color.parseColor("#1C1C1E")
    fun calloutBubble(ctx: Context)   = if (dark(ctx)) Color.parseColor("#4F4F53") else Color.parseColor("#333333")
    fun popup(ctx: Context)           = if (dark(ctx)) Color.parseColor("#414144") else Color.parseColor("#F5F5F5")
    fun popupText(ctx: Context)       = if (dark(ctx)) Color.WHITE                else Color.parseColor("#1C1C1E")
    fun selectedItem()                = Color.parseColor("#007AFF")

    private fun dark(ctx: Context) =
        ctx.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK == Configuration.UI_MODE_NIGHT_YES
}
